File: Dynamic_Programming/2786.Visit_Array_Positions_to_Maximize_Score/Python/2786.Visit_Array_Positions_to_Maximize_Score.py
class Solution(object):
    def maxScore(self, nums, x):
        """
        :type nums: List[int]
        :type x: int
        :rtype: int
        """
        #============================================#
        # Dynamic-programming based traversal method #
        #============================================#

        ############
        #Initialize
        ##### Length of nums array #####
        len_nums = len(nums)

        ##### Record dynamic-programming (i.e. odd, even) #####
        record_dp_odd = (nums[0] - (0 if (nums[0] % 2) else x))
        record_dp_even = (nums[0] - (0 if (not (nums[0] % 2)) else x))


        ##########################################
        #Dynamic-programming based loop traversal
        for nums_idx in range(1, len_nums):

            ##### Check if the current indexed-value is odd or even #####
            if ((nums[nums_idx] % 2)):
                record_dp_odd = max((record_dp_odd + nums[nums_idx]), ((record_dp_even + nums[nums_idx]) - x)) #Keep updating/overwriting
            else:
                record_dp_even = max((record_dp_even + nums[nums_idx]), ((record_dp_odd + nums[nums_idx]) - x)) #Keep updating/overwriting

        return max(record_dp_odd, record_dp_even)

File: Dynamic_Programming/2786.Visit_Array_Positions_to_Maximize_Score/C++/2786.Visit_Array_Positions_to_Maximize_Score.cpp
class Solution {
public:
    long long maxScore(vector<int> & nums, int x) 
    {
        //======================================//
        // Input type:                          //
        //   - vector<int> ref. (nums vector)   //
        //   - int (x score)                    //
        // Return type:                         //
        //   - long long (result maximun score) //
        //======================================//

        /*Initialize*/
        ///// Length of nums vector /////
        int len_nums = nums.size();

        ///// Record dynamic-programming (i.e. odd, even) /////
        long long record_dp_odd = (nums[0] - ((nums[0] % 2) ? 0 : x));
        long long record_dp_even = (nums[0] - ((! (nums[0] % 2)) ? 0 : x));


        /*Dynamic-programming based loop traversal*/
        for (int nums_idx = 1; (nums_idx < len_nums); (nums_idx++)) //Whole
        {
            ///// Check if the current indexed-value is odd or even /////
            if ((nums[nums_idx] % 2))
            {
                record_dp_odd = max((record_dp_odd + nums[nums_idx]), ((record_dp_even + nums[nums_idx]) - x)); //Keep updating, overwriting
            }
            else { record_dp_even = max((record_dp_even + nums[nums_idx]), ((record_dp_odd + nums[nums_idx]) - x)); } //Keep updating, overwriting

        } //Whole

        return max(record_dp_odd, record_dp_even);
    }
};
